### PR TITLE
[SPARK-49066][SQL][TESTS] Refactor `OrcEncryptionSuite` and make `spark.hadoop.hadoop.security.key.provider.path` effective only within `OrcEncryptionSuite`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr moves the global scope test configuration `spark.hadoop.hadoop.security.key.provider.path`, which is configured in the parent `pom.xml` and `SparkBuild.scala`, to `OrcEncryptionSuite` to ensure that it is effective only within `OrcEncryptionSuite`.

To achieve this, the pr also refactors `OrcEncryptionSuite`:
1. Overrides `beforeAll` to back up the contents of `CryptoUtils#keyProviderCache`.
2. Overrides `afterAll` to restore the contents of `CryptoUtils#keyProviderCache`.

This ensures that `CryptoUtils#keyProviderCache` is isolated during the test process of `OrcEncryptionSuite`.

### Why are the changes needed?
The test configuration `spark.hadoop.hadoop.security.key.provider.path` in the parent `pom.xml` and `SparkBuild.scala` is effective globally, which leads to the possibility that other Orc writing test cases, besides `OrcEncryptionSuite`, might also be affected by this configuration and use `test.org.apache.spark.sql.execution.datasources.orc.FakeKeyProvider.Factory`。

### Does this PR introduce _any_ user-facing change?
No, just for test.

### How was this patch tested?
Pass GitHub Actions

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -135,6 +135,12 @@ jobs:
         IMG_URL="ghcr.io/$REPO_OWNER/$IMG_NAME"
         echo "image_url=$IMG_URL" >> $GITHUB_OUTPUT
 
+  maven-test:
```
Comment (LuciferYang) [inline .github/workflows/build_and_test.yml:138]: will revert after maven test pass

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcEncryptionSuite.scala [hunk 2]
```diff
@@ -17,20 +17,53 @@
...
 class OrcEncryptionSuite extends OrcTest with SharedSparkSession {
   import testImplicits._
 
+  override def sparkConf: SparkConf = {
+    super.sparkConf.set("spark.hadoop.hadoop.security.key.provider.path", "test:///")
+  }
+
+  override def beforeAll(): Unit = {
+    // Backup `CryptoUtils#keyProviderCache` and clear it.
+    keyProviderCacheRef.entrySet()
```
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcEncryptionSuite.scala:41]: At this time, the content of `keyProviderCacheRef` might be 

```
hadoop.java.security.SecureRandom -> org.apache.orc.impl.NullKeyProvider
```.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcEncryptionSuite.scala [hunk 3]
```diff
@@ -17,20 +17,53 @@
...
+      .forEach(e => keyProviderCacheBackup.put(e.getKey, e.getValue))
+    keyProviderCacheRef.clear()
+    super.beforeAll()
+  }
+
+  override def afterAll(): Unit = {
+    super.afterAll()
+    // Restore `CryptoUtils#keyProviderCache`.
+    keyProviderCacheRef.clear()
```
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcEncryptionSuite.scala:50]: At this time, the content of `keyProviderCacheRef` might be

```
memory.java.security.SecureRandom -> org.apache.orc.InMemoryKeystore
hadoop.java.security.SecureRandom -> org.apache.orc.impl.KeyProviderImpl
```

**Comment by LuciferYang** [comment]
For other test cases, such as `OrcV1QuerySuite`, the code has been modified to print the contents of `CryptoUtils#keyProviderCache` in `afterAll`.

Before this PR, the contents of `CryptoUtils#keyProviderCache` were:

```
hadoop.java.security.SecureRandom -> org.apache.orc.impl.KeyProviderImpl@17fade83
```

After this PR, the contents of `CryptoUtils#keyProviderCache` are:

```
hadoop.java.security.SecureRandom -> org.apache.orc.impl.NullKeyProvider@54f3abc2
```

**Comment by LuciferYang** [comment]
cc @dongjoon-hyun I'm not sure if the original intention of adding `spark.hadoop.hadoop.security.key.provider.path` in `SparkBuild.scala` was to have it take effect globally, so please let me know if we do not need this pr. Thanks ~

**Comment by LuciferYang** [comment]
> To achieve this, the pr also refactors `OrcEncryptionSuite`:
> 1. Overrides `beforeAll` to back up the contents of `CryptoUtils#keyProviderCache`.
> 2. Overrides `afterAll` to restore the contents of `CryptoUtils#keyProviderCache`.
> 
> This ensures that `CryptoUtils#keyProviderCache` is isolated during the test process of `OrcEncryptionSuite`.

Without this change, `OrcEncryptionSuite` would fail because, after this PR, the implementation corresponding to key `hadoop.java.security.SecureRandom` in `CryptoUtils#keyProviderCache` has been designated as `org.apache.orc.impl.NullKeyProvider`

**Comment by dongjoon-hyun** [comment]
Thank you. Will take a look before the end of this week.

**Comment by JeonDaehong** [comment]
Hello, I am Mr. Jeon, a developer in Korea.
I would like to become a Committer for Apache Spark. Could you please give me some advice on how to achieve this?

**Comment by LuciferYang** [comment]
Thanks @dongjoon-hyun  ~

**Review by yaooqinn — APPROVED** [review]
LGTM from my side to minimize the influence of test configurations.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM, @LuciferYang . Please don't forget to remove `build_and_test.yml` change~

**Comment by LuciferYang** [comment]
> +1, LGTM, @LuciferYang . Please don't forget to remove `build_and_test.yml` change~

done :)

**Comment by dongjoon-hyun** [comment]
The failure of `SorterSuite` is a known flaky one.
```
[info] *** 1 SUITE ABORTED ***
[error] Error: Total 4117, Failed 0, Errors 1, Passed 4116, Ignored 8, Canceled 2
[error] Error during tests:
[error] 	org.apache.spark.util.collection.SorterSuite
[error] (core / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 3192 s (53:12), completed Jul 31, 2024, 6:12:53 PM
```

Let me merge this.

**Comment by dongjoon-hyun** [comment]
Merged to master/3.5.

**Comment by dongjoon-hyun** [comment]
This is reverted from branch-3.5 due to the compilation failure.

**Comment by LuciferYang** [comment]
OK, I will submit a PR for branch-3.5

**Pull request closed without merging**